### PR TITLE
Pm/charge deposition particle binning

### PR DESCRIPTION
**Pull request by PhilMiller**

### CODE REVIEW

File: Source/Particles/Deposition/ChargeDeposition.H [hunk 1]
```diff
@@ -205,19 +250,19 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
...
             }
 
-            // Deposit charge into rho_tmp
+            // Deposit charge into buf
 #if (defined WARPX_DIM_XZ) || (defined WARPX_DIM_RZ)
             for (int iz=0; iz<=depos_order; iz++){
                 for (int ix=0; ix<=depos_order; ix++){
                     amrex::Gpu::Atomic::AddNoRet(
-                        &rho_tmp(lo.x+i+ix, lo.y+k+iz, 0, 0),
+                        &buf(lo.x+i+ix, lo.y+k+iz, 0, 0),
```
Comment (PhilMiller) [inline Source/Particles/Deposition/ChargeDeposition.H:272]: I'm getting out of bounds accesses here, by up to a few points, positive or negative, in both X and Z. E.g.

` (-10,5,0,0) is out of bound (-2:130,-2:10,0:0,0:0)`
` (29,8,0,0) is out of bound (27:32,-2:3,0:0,0:0)`

There are hundreds of these in the first call to `depositChargeShapeN`. The logic all looks reasonable to me, but there must be some subtlety in the box sizing calculations that I'm missing.

File: Source/Particles/Deposition/ChargeDeposition.H [hunk 2]
```diff
@@ -90,22 +114,31 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
...
     const std::size_t max_shared_mem_bytes = amrex::Gpu::Device::sharedMemPerBlock();
 
     if (shared_mem_bytes > max_shared_mem_bytes) {
       std::cout << "Tile size for charge deposition of " << shared_mem_bytes
                 << " bytes would not fit in GPU's shared memory of " << max_shared_mem_bytes << " bytes" << std::endl;
-#if 1
-      std::cout << "Falling back to direct global memory access" << std::endl;
-      shared_mem_bytes = 0;
-#else
       amrex::Abort("Tile size too big for GPU shared memory charge deposition");
```
Comment (PhilMiller) [inline Source/Particles/Deposition/ChargeDeposition.H:143]: This case gets hit at the first step after initialization, with obviously garbage huge values:

`Tile size for charge deposition of 52377931632 bytes would not fit in GPU's shared memory of 49152 bytes`

**Comment by PhilMiller** [comment]
@atmyers Would you mind having a look at the out of bounds access I'm running into?

File: Source/Particles/Deposition/ChargeDeposition.H [hunk 3]
```diff
@@ -90,22 +114,31 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
...
 #endif
 
+    const auto dxiarr = geom.InvCellSizeArray();
+    const auto plo = geom.ProbLoArray();
+    const auto domain = geom.Domain();
+
+    const amrex::Box sample_tbox(IntVect(AMREX_D_DECL(0, 0, 0)), a_tbox_max_size - 1);
+    amrex::Box sample_tbox_x = convert(sample_tbox, ix_type);
+
+    sample_tbox_x.grow(depos_order);
```
Comment (PhilMiller) [inline Source/Particles/Deposition/ChargeDeposition.H:127]: For the out of bounds access, I tried adding a margin of up to `+8` here and in the corresponding `.grow()` call in the kernel, but it still fails on the negative values.

**Comment by PhilMiller** [comment]
I think the uses of `xmin/ymin/zmin` need to be corrected to be relative to the individual tile, and not the larger box that it lives in or the overall domain.

File: Source/Particles/Deposition/ChargeDeposition.H (hunk 2 above)
Comment (PhilMiller) [reply Source/Particles/Deposition/ChargeDeposition.H:143]: Andrew's fixes handled this.

File: Source/Particles/Deposition/ChargeDeposition.H (hunk 3 above)
Comment (PhilMiller) [reply Source/Particles/Deposition/ChargeDeposition.H:127]: Andrew's fixes handled this

**Comment by PhilMiller** [comment]
Next big challenge is to make this compile and run, correctly, on non-GPU targets as well.

**Comment by PhilMiller** [comment]
Posted upstream as https://github.com/ECP-WarpX/WarpX/pull/2550

Hopefully it can get merged there quickly enough that we don't feel motivated to merge this PR.

File: Source/Particles/Deposition/ChargeDeposition.H [hunk 4]
```diff
@@ -118,28 +154,51 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
...
+          ParticleReal xp, yp, zp;
+          GetPosition(bin_start, xp, yp, zp);
+#if (AMREX_SPACEDIM == 3)
+          IntVect iv = IntVect(int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((yp-plo[1])*dxiarr[1])),
+                               int(amrex::Math::floor((zp-plo[2])*dxiarr[2])));
+#elif (AMREX_SPACEDIM == 2)
+          IntVect iv = IntVect(
+                               int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((zp-plo[1])*dxiarr[1])));
```
Comment (KZhu-ME) [inline Source/Particles/Deposition/ChargeDeposition.H:174]: The 1D definition for `iv` was missing.
```suggestion
                               int(amrex::Math::floor((zp-plo[1])*dxiarr[1])));
#elif (AMREX_SPACEDIM == 1)
          IntVect iv = IntVect(int(amrex::math::floor((zp-plo[0])*dxiarr[0])));
                              
```

**Pull request closed without merging**